### PR TITLE
Initial migration to github

### PR DESCRIPTION
**Pull request by zwall-bp**
The current state of this cookbook is deployable, but would improve with some touchups and cleaning.

## osl_element
This resource is fairly simple, it deploys an instance of Element, a Matrix webapp client, on port 8000. This relies on another resource or recipe to then put it behind a reverse proxy for security.

## osl_synapse
This resource deployed the actual synapse server, using Docker. When deploying with other Matrix app services, this must be deployed first, in order to set up the environment.

It's possible to connect to a Postgres DB or using sqlite locally. Only use sqlite for testing. 

## osl_heisenbridge
This resource adds in the heisenbridge app service, adding a bot to the Synapse server which allows us to bridge IRC chats into a Matrix room. Managed with Docker.

## osl_hookshot
A simple webhook service for Matrix, deployed using Docker.

TODO services:
* MS Teams bridge (Turns out that the only, officialy available bridge is behind an enterprise paywall)

### CODE REVIEW

File: files/chat.osuosl.org.signing.key [hunk 1]
```diff
@@ -0,0 +1 @@
+ed25519 a_cDWo IgroQ8oS8U+uxtDmS8czW3fVnTO/qmdJL7EOboYz4Os
```
Comment (ramereth) [inline files/chat.osuosl.org.signing.key]: Shouldn't this be a secret?

File: files/element-config.json [hunk 2]
```diff
@@ -0,0 +1,49 @@
+{
+    "default_server_config": {
+        "m.homeserver": {
+            "base_url": "https://talk.osuosl.intnet",
+            "server_name": "talk.osuosl.intnet"
```
Comment (ramereth) [inline files/element-config.json]: Shouldn't these be configurable via a template?

File: files/homeserver.yaml [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+registration_shared_secret: "S&m5Sm@YjO98DU^#cpNtI;z1toN6_@w8AWtDd8JbCsdvV*mgD_"
+report_stats: false
+macaroon_secret_key: "k8hF=4hY^moi:P_Nimook9xhw*HKN2Qzs*1tD@-1QfQbr*Av#L"
+form_secret: "&DLY6FktADN7sow+&Ge;d6#WDU-s9zcj*t3Irqd0XE;D4MaLOH"
+signing_key_path: "/data/chat.osuosl.org.signing.key"
+trusted_key_servers:
+  - server_name: "matrix.org"
+
+
+# vim:ft=yaml
```
Comment (ramereth) [inline files/homeserver.yaml]: This should all be in a template instead with the secrets being sensitive.

File: kitchen.yml [hunk 4]
```diff
@@ -7,8 +7,19 @@ provisioner:
...
 
 suites:
-  - name: default
+  - name: synapse
     run_list:
-      - recipe[osl-matrix::default]
+      - recipe[osl-matrix-test::synapse]
+    driver:
+      name: openstack
+      flavor_ref: 'm1.medium'
```
Comment (ramereth) [inline kitchen.yml]: I believe this is the global default now anyway so you can remove this.

File: libraries/helpers.rb [hunk 5]
```diff
@@ -0,0 +1,29 @@
+module OSLMatrix
+  module Cookbook
+    module Helpers
+
+      # Create the synapse user, and the recommended directories
+      def init_environment()
```
Comment (ramereth) [inline libraries/helpers.rb]: Please prepend method names with `osl_matrix_` so we don't run into conflicts elsewhere.

File: metadata.rb [hunk 6]
```diff
@@ -2,10 +2,16 @@
 maintainer        'Oregon State University'
 maintainer_email  'chef@osuosl.org'
 license           'All Rights Reserved'
-description       'Installs/Configures osl-matrix'
+description       'Creates and manages a matrix synapse service for the Open Source Lab'
 issues_url        'https://github.com/osuosl-cookbooks/osl-matrix/issues'
 source_url        'https://github.com/osuosl-cookbooks/osl-matrix'
-chef_version      '>= 17.0'
+chef_version      '>= 16.0'
```
Comment (ramereth) [inline metadata.rb]: Please keep this on 17.

File: metadata.rb [hunk 7]
```diff
@@ -2,10 +2,16 @@
...
 version           '0.1.0'
 
 supports          'almalinux', '~> 8.0'
+
+depends           'docker'
```
Comment (ramereth) [inline metadata.rb]: No need to directly depend on `docker` since you're pulling in `osl-docker`

File: recipes/default.rb [hunk 8]
```diff
@@ -15,3 +15,30 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
+
+include_recipe 'osl-docker'
+
+# Synapse management user
+user 'synapse-host' do
```
Comment (ramereth) [inline recipes/default.rb]: Any reason why this can't be the following?
```suggestion
user 'synapse' do
```

File: recipes/default.rb [hunk 9]
```diff
@@ -15,3 +15,30 @@
...
+# Synapse configuration, used within the docker container as a volume
+directory '/etc/synapse' do
+  owner 'synapse-host'
+  group 'synapse-host'
+  mode '750'
+end
+
+cookbook_file '/etc/synapse/homeserver.yaml' do
+  source 'homeserver.yaml'
+
```
Comment (ramereth) [inline recipes/default.rb]: If the filename name are the same, you can leave this out. Plus remove the extra space.

File: recipes/default.rb [hunk 10]
```diff
@@ -15,3 +15,30 @@
...
+  owner 'synapse-host'
+  group 'synapse-host'
+end
+
+docker_container 'Matrix Synapse' do
+  repo 'matrixdotorg/synapse'
+  volumes ['/etc/synapse:/data']
+  env ["UID=#{node['ect']['passwd']['synapse-host']['uid']}", "GID=#{node['ect']['passwd']['synapse-host']['gid']}"]
```
Comment (ramereth) [inline recipes/default.rb]: Please expand this array out i.e.:
```suggestion
  env [
    "UID=#{node['ect']['passwd']['synapse-host']['uid']}",
    "GID=#{node['ect']['passwd']['synapse-host']['gid']}"
  ]
```
Also, why the use of attributes here?

File: resources/osl_element.rb [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+property :domain, String, name_property: true
+property :matrix_domain, String
+
+action :create do
+  include_recipe 'osl-docker'
+  include_recipe 'osl-nginx'
+
+  hostname node['hostname'] do
+    aliases [new_resource.domain]
+  end
```
Comment (ramereth) [inline resources/osl_element.rb]: Why is this needed?

File: resources/osl_element.rb [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+
+  directory '/srv/element'
```
Comment (ramereth) [inline resources/osl_element.rb]: I'd prefer to put it under `/opt`
```suggestion
  directory '/opt/element'
```

File: resources/osl_element.rb [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+# This resource allows for the deployment of Element, a Matrix client, as a website.
+
+resource_name :osl_element
+provides :osl_element
+unified_mode true
+
+default_action :create
+
+property :domain, String, name_property: true
+property :matrix_domain, String
```
Comment (ramereth) [inline resources/osl_element.rb]: Where does this end up being used?

File: resources/osl_heisenbridge.rb [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+#
+# Heisenbridge is an IRC bouncer for a Matrix server
+
+resource_name :osl_heisenbridge
+provides :osl_heisenbridge
+unified_mode true
+
+default_action :create
+
+property :container_name, String, name_property: true
```
Comment (ramereth) [inline resources/osl_heisenbridge.rb]: Not sure this is the best name based on what I see below.

File: resources/osl_heisenbridge.rb [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+
+action :create do
+  # Pull down the heisenbridge image
+  docker_image 'hif1/heisenbridge'
+  
+  # Generate the app service file
+  execute 'heisenbridge app service' do
+    command "docker run --volume=/srv/synapse:/data hif1/heisenbridge  'python -m heisenbridge' -c /data/#{new_resource.container_name}.yaml --generate --listen-address #{new_resource.container_name}"
+    creates "/srv/synapse/#{new_resource.container_name}.yaml"
+  end
```
Comment (ramereth) [inline resources/osl_heisenbridge.rb]: Is there not another way of doing this? This is pretty hacky.

File: resources/osl_heisenbridge.rb [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+
+  # Update the file's permissions for synapse-host
+  file 'update heisenbridge perms' do
+    file "/srv/synapse/#{new_resource.container_name}.yaml"
```
Comment (ramereth) [inline resources/osl_heisenbridge.rb]: Just use the filename as the name property please.

File: resources/osl_heisenbridge.rb [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+    mode '644'
+  end
+
+  # Create the docker container and queue it for a delayed deployment
+  docker_container new_resource.container_name do
+    repo 'hif1/heisenbridge'
+    volumes ["/srv/synapse/#{new_resource.container_name}.yaml:/data/#{new_resource.container_name}.yaml"]
+    entrypoint "python -m heisenbridge -c /data/#{new_resource.container_name}.yaml http://osl-matrix-synapse:8008"
```
Comment (ramereth) [inline resources/osl_heisenbridge.rb]: The hostname at the end should be a property of this resource so it's configurable.

File: resources/osl_matrix.rb [hunk 18]
```diff
@@ -0,0 +1,92 @@
+resource_name :osl_matrix
```
Comment (ramereth) [inline resources/osl_matrix.rb]: Let's go ahead and rename this as `osl_synapse` since that's what we're actually deploying. The cookbook deals with all things Matrix and we may end up deploying another server down the road so this will keep it separate.

File: resources/osl_matrix.rb [hunk 19]
```diff
@@ -0,0 +1,92 @@
...
+    owner 'synapse-host'
+    mode '700'
+  end
+
+  # Create the docker network for Synapse-related items
+  docker_network 'synapse-network'
+
+  # Generate secret keys for Synapse. Only generate once.
+  file '/srv/synapse/keys/registration.key' do
+    content SecureRandom.base64(32)
```
Comment (ramereth) [inline resources/osl_matrix.rb]: Please convert this to a simple helper method and remove the require above which will get added to the helper instead.

File: resources/osl_matrix.rb [hunk 20]
```diff
@@ -0,0 +1,92 @@
...
+      pg_host: new_resource.pg_host,
+      pg_name: new_resource.pg_name,
+      pg_username: new_resource.pg_username,
+      pg_password: new_resource.pg_password,
+      sqlite: new_resource.use_sqlite,
+      user_config: new_resource.config,
+    )
+  end
+
+  nSynapseUID = shell_out('id -u synapse-host').stdout
```
Comment (ramereth) [inline resources/osl_matrix.rb]: Instead of doing this, use the [ohai](https://docs.chef.io/resources/ohai/#examples) resource to reload. See the example I linked there. Then you can directly get the UID without doing a shellout.

File: resources/osl_matrix.rb [hunk 21]
```diff
@@ -0,0 +1,92 @@
...
+
+  docker_image 'matrixdotorg/synapse'
+
+  docker_container 'osl-matrix-synapse' do
```
Comment (ramereth) [inline resources/osl_matrix.rb]: This should probably use the name resource as a part of this name. Maybe something like the following:
```suggestion
  docker_container 'matrix-synapse-#{new_resource.name}' do
```

File: spec/spec_helper.rb [hunk 22]
```diff
@@ -1,6 +1,21 @@
 require 'chefspec'
 require 'chefspec/berkshelf'
 
+CENTOS_7 = {
+  platform: 'centos',
+  version: '7',
+}.freeze
+
```
Comment (ramereth) [inline spec/spec_helper.rb]: Please remove CentOS 7 as we're not going to worry about that. We should remove that from our generator template.

File: test/cookbooks/osl-matrix-test/recipes/synapse.rb [hunk 23]
```diff
@@ -0,0 +1,23 @@
+include_recipe 'osl-docker'
```
Comment (ramereth) [inline test/cookbooks/osl-matrix-test/recipes/synapse.rb]: This should technically not be needed if you're doing this inside the resources.

**Review by ramereth — COMMENTED** [review]
Some initial comments, certainly not complete but should be enough to get you started. Great work!

File: files/chat.osuosl.org.signing.key (hunk 1 above)
Comment (zwall-bp) [reply files/chat.osuosl.org.signing.key]: I didn't intend on committing anything that was present in the `files/` directory, as it was only for testing purposes. The cookbook doesn't rely on any of these files.

File: libraries/helpers.rb (hunk 5 above)
Comment (zwall-bp) [reply libraries/helpers.rb:6]: The helpers function was for a previous implementation, and is no longer required. I've removed the file.

File: resources/osl_heisenbridge.rb (hunk 15 above)
Comment (zwall-bp) [reply resources/osl_heisenbridge.rb]: This is Heisenbridge's [recommended way of generating an appservice file](https://github.com/hifi/heisenbridge/blob/master/docker-compose/init/heisenbridge-init.sh).

I was considering using a template instead, but decided to use this in the event Heisenbridge updates what is in the file, with this allowing us to easily adapt to the new content without needing to upgrade the cookbook.

Since Appservice files are usually [similar to each other](https://matrix-org.github.io/synapse/latest/application_services.html), the next build will instead be using a generic template instead.

File: resources/osl_heisenbridge.rb (hunk 14 above)
Comment (zwall-bp) [reply resources/osl_heisenbridge.rb]: In order to make linking App Services, like Heisenbridge, to the main Synapse server easy for the developer. It's best if everything related to the container is named the exact same.

I have experimented with instead updating the App Service list using attributes inside of the resources, but removed it as it goes against the spirit of resources. (See commented out code in osl_heisenbridge)

File: resources/osl_matrix.rb [hunk 24]
```diff
@@ -0,0 +1,92 @@
...
+    repo 'matrixdotorg/synapse'
+    port ['8008:8008']
+    volumes ['/srv/synapse/:/data']
+    env ["UID=#{nSynapseUID}"]
```
Comment (ramereth) [inline resources/osl_matrix.rb]: You can replace what you have above with the following I believe:
```suggestion
    env ["UID=#{Etc.getpwnam('synapse-host').uid.to_s}"]
```

**Comment by zwall-bp** [comment]
Along with all of the requested changes, the biggest update is that I have solved the chicken and egg problem, by restarting the docker containers when they fail.

While I was hoping to find a more elegant solution, this is the best one, unless we want to work with Chef on a lower level, messing with the run order at converge time.

**Comment by zwall-bp** [comment]
Synapse is now in a state where we can start using it for basic communication, along with also having the ability to host meetings w/ screen sharing. My next goal is to add in a [Hookshot](https://github.com/matrix-org/matrix-hookshot) docker as a resource, with it being another appservice.

The provided `osl_heisenbridge` gives a good example for how future appservice resources should be set up.

File: files/chat.osuosl.org.log.config [hunk 25]
```diff
@@ -0,0 +1,39 @@
...
+    
+
+root:
+    level: INFO
+
+
+    handlers: [console]
+
+
+disable_existing_loggers: false
```
Comment (ramereth) [inline files/chat.osuosl.org.log.config]: Is this static file needed? The name is rather specific it seems.

File: files/chat.osuosl.org.signing.key (hunk 1 above)
Comment (ramereth) [reply files/chat.osuosl.org.signing.key]: Can this be removed then?

File: files/element-config.json (hunk 2 above)
Comment (ramereth) [reply files/element-config.json]: Same question :point_up:

File: files/homeserver.yaml [hunk 26]
```diff
@@ -0,0 +1,37 @@
...
+    database: /data/homeserver.db
+log_config: "/data/chat.osuosl.org.log.config"
+media_store_path: /data/media_store
+registration_shared_secret: "S&m5Sm@YjO98DU^#cpNtI;z1toN6_@w8AWtDd8JbCsdvV*mgD_"
+report_stats: false
+macaroon_secret_key: "k8hF=4hY^moi:P_Nimook9xhw*HKN2Qzs*1tD@-1QfQbr*Av#L"
+form_secret: "&DLY6FktADN7sow+&Ge;d6#WDU-s9zcj*t3Irqd0XE;D4MaLOH"
+signing_key_path: "/data/chat.osuosl.org.signing.key"
+trusted_key_servers:
+  - server_name: "matrix.org"
```
Comment (ramereth) [inline files/homeserver.yaml]: Why isn't this a template?

File: kitchen.yml (hunk 4 above)
Comment (ramereth) [reply kitchen.yml]: This is still the case :point_up:

File: resources/osl_heisenbridge.rb [hunk 27]
```diff
@@ -0,0 +1,61 @@
...
+property :port, Integer, default: 9898
+property :matrix_host_resource, Symbol, required: true, default: :osl_synapse
+property :matrix_host_domain, String, required: true
+
+action :create do
+  # Get the required information taken from the matrix host resource
+  matrix_resource = find_resource(new_resource.matrix_host_resource, new_resource.matrix_host_domain)
+  path = matrix_resource.path
+  name = matrix_resource.name
+  network = matrix_resource.network
```
Comment (ramereth) [inline resources/osl_heisenbridge.rb]: I don't think this should be automated like this. This is bound to create issues later. When we deploy this in production, we will know all of these so we should just have properties in this resource to set it.

File: resources/osl_synapse.rb [hunk 28]
```diff
@@ -0,0 +1,95 @@
...
+property :tag, String, default: 'latest'
+property :use_sqlite, [true, false], default: false
+
+action :create do
+  # Check to see if a name was given
+  new_resource.name = "matrix-synapse-#{new_resource.domain}" unless new_resource.name
+  # Get the path to use for this resource
+  new_resource.path = "/opt/synapse-#{new_resource.name}" unless new_resource.path
+  # Set the docker network
+  new_resource.network = "synapse-network-#{new_resource.name}" unless new_resource.network
```
Comment (ramereth) [inline resources/osl_synapse.rb]: These should all be defaults above using `lazy` to set them based on the `domain` property.

File: resources/osl_synapse.rb [hunk 29]
```diff
@@ -0,0 +1,95 @@
...
+
+  # Create the docker network for Synapse-related items
+  docker_network new_resource.network
+
+  # Generate secret keys for Synapse. Only generate once.
+  file "#{new_resource.path}/keys/registration.key" do
+    content osl_matrix_genkey
+    owner 'synapse'
+    mode '400'
+
```
Comment (ramereth) [inline resources/osl_synapse.rb]: Please remove this extra new line. Also make this a sensitive resource.

File: resources/osl_synapse.rb [hunk 30]
```diff
@@ -0,0 +1,95 @@
...
+    action :create_if_missing
+  end
+
+  # Generate homeserver.yaml
+  template "#{new_resource.path}/homeserver.yaml" do
+    source 'homeserver.yaml.erb'
+    cookbook 'osl-matrix'
+    owner 'synapse'
+    mode '600'
```
Comment (ramereth) [inline resources/osl_synapse.rb]: Please make this a sensitive resource.

File: resources/osl_heisenbridge.rb [hunk 31]
```diff
@@ -0,0 +1,61 @@
...
+
+  # Pull down the heisenbridge image
+  docker_image 'hif1/heisenbridge'
+
+  # Generate the app service file
+  template "#{path}/#{new_resource.container_name}.yaml" do
+    source 'appservice.erb'
+    cookbook 'osl-matrix'
+    mode '644'
```
Comment (ramereth) [inline resources/osl_heisenbridge.rb]: Please make this a sensitive resource.

File: resources/osl_element.rb [hunk 32]
```diff
@@ -0,0 +1,31 @@
...
+
+action :create do
+  include_recipe 'osl-docker'
+
+  directory '/opt/element'
+
+  template '/opt/element/config.json' do
+    source 'element-config.json.erb'
+    cookbook 'osl-matrix'
+    variables(fqdn: new_resource.matrix_domain)
```
Comment (ramereth) [inline resources/osl_element.rb]: Please make this a sensitive resource.

File: spec/spec_helper.rb [hunk 33]
```diff
@@ -1,6 +1,21 @@
...
 
+ALMALINUX_8 = {
+  platform: 'almalinux',
+  version: '8',
+}.freeze
+
+CENTOS_8 = {
+  platform: 'centos',
+  version: '8',
+}.freeze
```
Comment (ramereth) [inline spec/spec_helper.rb]: We're not testing on this platform, please remove.

File: spec/unit/resources/osl_element.rb [hunk 34]
```diff
@@ -0,0 +1,25 @@
...
+# Spec:: osl_element
+#
+# Copyright:: 2023, Oregon State University
+#
+
+require_relative '../../spec_helper'
+
+describe 'osl-matrix-test::element' do
+  cached(:subject) { chef_run }
+  platform 'almalinux', '9'
```
Comment (ramereth) [inline spec/unit/resources/osl_element.rb]: We're not running on 9 yet, please change this to 8.

File: test/cookbooks/osl-matrix-test/recipes/synapse.rb [hunk 35]
```diff
@@ -0,0 +1,25 @@
...
+    config:
+      enabled: true
+      uri: "ldap://ldap.osuosl.org:389"
+      start_tls: true
+      base: "ou=People,dc=osuosl,dc=org"
+      attributes:
+        uid: "uid"
+        mail: "mail"
+        name: "givenName"
+  EOF
```
Comment (ramereth) [inline test/cookbooks/osl-matrix-test/recipes/synapse.rb]: Why can't this be given as a Hash since it just seems to be? If this needs to be converted to yaml, there is a ruby library which will do that for you.

File: test/cookbooks/osl-matrix-test/recipes/element.rb [hunk 36]
```diff
@@ -0,0 +1 @@
+osl_element 'talk.osuosl.intnet'
```
Comment (ramereth) [inline test/cookbooks/osl-matrix-test/recipes/element.rb]: Please use `example.org` as the domain for testing all of these.
```suggestion
osl_element 'talk.example.org'
```

**Review by ramereth — CHANGES_REQUESTED** [review]

File: files/chat.osuosl.org.signing.key (hunk 1 above)
Comment (zwall-bp) [reply files/chat.osuosl.org.signing.key]: It was an oversight.

**Comment by ramereth** [comment]
@zwall-bp please remind me the status on this?

File: libraries/helpers.rb [hunk 37]
```diff
@@ -0,0 +1,19 @@
...
+  module Cookbook
+    module Helpers
+      # Generate a secret key. Mainly used when wanting a registration key
+      def osl_matrix_genkey(len = 64)
+        require 'securerandom'
+        SecureRandom.base64(len)
+      end
+
+      # Get the name of the synapse docker container name, given the name of the synapse resource which creates it
+      def osl_synapse_docker_name(strSynapseResource)
```
Comment (ramereth) [inline libraries/helpers.rb]: I prefer names like this please:
```suggestion
      def osl_synapse_docker_name(synapse_resource)
```

File: resources/osl_element.rb [hunk 38]
```diff
@@ -0,0 +1,32 @@
...
+    sensitive true
```
Comment (ramereth) [inline resources/osl_element.rb]: This probably should notify restarting the docker_container below when this gets updated.

File: kitchen.yml (hunk 4 above)
Comment (ramereth) [reply kitchen.yml]: Please address this.

File: kitchen.yml [hunk 39]
```diff
@@ -7,8 +7,18 @@ provisioner:
...
   multiple_converge: 2
   deprecations_as_errors: true
+  data_bags_path: test/integration/data_bags
 
 suites:
-  - name: default
+  - name: synapse
     run_list:
-      - recipe[osl-matrix::default]
+      - recipe[osl-matrix-test::synapse]
```
Comment (ramereth) [inline kitchen.yml:26]: This suite needs to run three times for it to finish properly. Please adjust it so that's the case.

File: resources/osl_heisenbridge.rb [hunk 40]
```diff
@@ -0,0 +1,59 @@
...
+      namespaces: {
+        users: [
+          {
+            exclusive: true,
+            regex: '\'@irc_.*\'',
+          },
+        ],
+      }
+    )
+    sensitive true
```
Comment (ramereth) [inline resources/osl_heisenbridge.rb]: This should notify a restart of the container when updated.

File: templates/element-config.json.erb [hunk 41]
```diff
@@ -0,0 +1,49 @@
...
+    },
+    "jitsi": {
+        "preferred_domain": "meet.element.io"
+    },
+    "element_call": {
+        "url": "https://call.element.io",
+        "participant_limit": 8,
+        "brand": "Element Call"
+    },
+    "map_style_url": "https://api.maptiler.com/maps/streets/style.json?key=fU3vlMsMn4Jb6dnEIFsx"
```
Comment (ramereth) [inline templates/element-config.json.erb:48]: Should this key be a secret?

File: test/cookbooks/osl-matrix-test/recipes/element.rb (hunk 36 above)
Comment (ramereth) [reply test/cookbooks/osl-matrix-test/recipes/element.rb]: Please fix this.

File: test/integration/default/inspec/default_spec.rb [hunk 42]
```diff
@@ -0,0 +1,4 @@
+# InSpec test for recipe osl-matrix::default
+
+# The InSpec reference, with examples and extensive documentation, can be
+# found at https://docs.chef.io/inspec/resources/
```
Comment (ramereth) [inline test/integration/default/inspec/default_spec.rb]: Please remove this.

File: test/integration/element/inspec/element_spec.rb [hunk 43]
```diff
@@ -0,0 +1,29 @@
...
+  it { should exist }
+  its('owner') { should eq 'synapse' }
+  its('mode') { should eq '400' }
+end
+
+describe file('/opt/synapse-chat.example.org/homeserver.yaml') do
+  it { should exist }
+  its('owner') { should eq 'synapse' }
+  its('mode') { should eq '600' }
+end
```
Comment (ramereth) [inline test/integration/element/inspec/element_spec.rb:29]: Can you please add some simple http checks to ensure the service is working?

**Review by ramereth — CHANGES_REQUESTED** [review]
Please also add tests for the synapse suite.

File: resources/osl_element.rb (hunk 38 above)
Comment (zwall-bp) [reply resources/osl_element.rb]: This luckily will not be needed as it's a static site. These configurations are loaded on the browser's end, when the user goes onto the website.

The docker container should already "update" the file's contents without needing to restart.

File: kitchen.yml (hunk 39 above)
Comment (zwall-bp) [reply kitchen.yml:26]: This error has been addressed to fix all errors before the end of the first converge.

**Comment by zwall-bp** [comment]
Still working on the requested changes, pushing the latest changes before clocking out.

**Comment by zwall-bp** [comment]
There is an issue right now with the firewall. At the end of the converge, IPTables restarts. This leads to the removal of rules for any running Docker containers.

File: resources/osl_element.rb (hunk 32 above)
Comment (ramereth) [reply resources/osl_element.rb]: This still needs to be a sensitive resource.

File: resources/osl_element.rb [hunk 44]
```diff
@@ -0,0 +1,31 @@
...
+
+  directory '/opt/element'
+
+  template '/opt/element/config.json' do
+    source 'element-config.json.erb'
+    cookbook 'osl-matrix'
+    variables(fqdn: new_resource.matrix_domain)
+  end
+
+  docker_image 'vectorim/element-web'
```
Comment (ramereth) [inline resources/osl_element.rb]: I believe this should restart the container below when it gets updated.

File: resources/osl_heisenbridge.rb [hunk 45]
```diff
@@ -0,0 +1,56 @@
...
+property :host_name, String, default: lazy { "matrix-synapse-#{host_domain}" }
+property :host_path, String, default: lazy { "/opt/synapse-#{host_name}" }
+property :host_network, String, default: lazy { "synapse-network-#{host_name}" }
+property :key_appservice, String, default: lazy { osl_matrix_genkey(host_name + container_name) }
+property :key_homeserver, String, default: lazy { osl_matrix_genkey(host_network + container_name) }
+
+action :create do
+  new_resource.host_name = new_resource.host_name
+  # Pull down the heisenbridge image
+  docker_image 'hif1/heisenbridge'
```
Comment (ramereth) [inline resources/osl_heisenbridge.rb]: I believe this should restart the container below when it gets updated.

File: resources/osl_hookshot.rb [hunk 46]
```diff
@@ -0,0 +1,157 @@
...
+property :host_domain, String, required: true
+property :host_name, String, default: lazy { "matrix-synapse-#{host_domain}" }
+property :host_path, String, default: lazy { "/opt/synapse-#{host_name}" }
+property :host_network, String, default: lazy { "synapse-network-#{host_name}" }
+property :key_appservice, String, default: lazy { osl_matrix_genkey(host_name + container_name) }
+property :key_homeserver, String, default: lazy { osl_matrix_genkey(host_network + container_name) }
+
+action :create do
+  # Pull down the Hookshot image
+  docker_image 'halfshot/matrix-hookshot'
```
Comment (ramereth) [inline resources/osl_hookshot.rb]: I believe this should restart the container below when it gets updated.

File: resources/osl_hookshot.rb [hunk 47]
```diff
@@ -0,0 +1,157 @@
...
+    'hookshot',
+    {
+      users: arrAppServiceUsers,
+    }
+  )
+
+  # Auto configuration for the homeserver, ignore if already set
+  # Passkey settings
+  # Not usually needed to be modified by user, as we generate our own, but just in case.
+  new_resource.config.merge!({
```
Comment (ramereth) [inline resources/osl_hookshot.rb]: You shouldn't be directly modifying the parameters in `new_resource` Can you please make a local variable to do this with?

File: resources/osl_synapse.rb [hunk 48]
```diff
@@ -0,0 +1,101 @@
...
+  # Generate homeserver.yaml
+  file "#{new_resource.path}/homeserver.yaml" do
+    content YAML.dump(new_resource.config).lines[1..-1].join
+    owner 'synapse'
+    group 'synapse'
+    mode '400'
+    sensitive true
+  end
+
+  docker_image 'matrixdotorg/synapse'
```
Comment (ramereth) [inline resources/osl_synapse.rb]: I believe this should restart the container below when it gets updated.

File: resources/osl_synapse.rb [hunk 49]
```diff
@@ -0,0 +1,101 @@
...
+    action :connect
+  end
+end
+
+# Restart the synapse container.
+action :restart do
+  docker_container new_resource.container_name do
+    action :restart
+  end
+end
```
Comment (ramereth) [inline resources/osl_synapse.rb]: Is this required externally from the resource?

**Comment by ramereth** [comment]
> There is an issue right now with the firewall. At the end of the converge, IPTables restarts. This leads to the removal of rules for any running Docker containers.

Is this related to the issue I'm seeing on the first run where I get the following error?
```
Docker::Error::TimeoutError
---------------------------
Container osl-hookshot-webhook failed to change to running state after 20 seconds
```

**Review by ramereth — CHANGES_REQUESTED** [review]
I'm also seeing the following errors with the synapse suite:
```
  Docker Container osl-irc-bridge
     ×  is expected to be running
     expected that `Docker Container osl-irc-bridge` is running

  Docker Container osl-hookshot-webhook
     ×  is expected to be running
     expected that `Docker Container osl-hookshot-webhook` is running
     ×  ports is expected to match "9000-9002->9000-9002/tcp"
     expected "" to match "9000-9002->9000-9002/tcp"
```

Not sure if this is also related to the firewall issues as well. Let me know.

File: resources/osl_synapse.rb (hunk 49 above)
Comment (zwall-bp) [reply resources/osl_synapse.rb]: It was used for testing, removed.

**Comment by ramereth** [comment]
!bump major ~

**Comment by osuosl-manatee** [comment]
Jenkins has merged this PR into `master` and has automatically performed a major-level version bump to v1.0.0.  Have a nice day!
